### DISCUSSION TITLE
Research & Develop on Evernode - Wallet Naming Service

### DISCUSSION THREAD

**Discussion by wojake** [XRPLBounties/Proposals#63] [Ideas]
# Wallet Naming Service

**Category:** Smart Contract Research & Development
**Tech:** Evernode; HotPocket consensus engine; the XRPL
**Summary**: This bounty is a Research and Development initiative on Evernode and web3. Evernode is a smart contract solution atop the XRP Ledger utilizing [the Hooks amendment](https://hooks.xrpl.org/) & [HotPocket](https://evernode.files.wordpress.com/2021/08/hotpocket-white-paper.pdf).

## Overview
Smart contracts have not come to fruition on the XRPL ecosystem yet but multiple research & development initiatives are being done to implement and adopt smart contract capabilities on the XRPL ecosystem without eliminating its current advantages. 

To name them, they are: (1) The Hooks amendment by XRPL Labs, (2) Evernode by Evernode Foundation, and (3) The EVM sidechain by Peersyst.

I'm proposing a bounty to research and develop on Evernode's technology, HotPocket, by developing a Wallet Naming Service for XRPL accounts. As Evernode & the Hooks amendment haven't been implemented on the mainnet, this development must be done and tested on a testnet.

# Bounty
Since this is a Research & Development bounty, it is mandatory for the applicant to explain the HotPocket consensus engine, Evernode and how their proposed Wallet Naming Service works under the hood on their service's HotPocket cluster (a cluster is a network of nodes). The Wallet Naming Service's code should be stored on a public (open source) GitHub repository.

### Governance Model
To further research Evernode Applications' capabilities, it is also complimentary to this initiative if a governance system is to be developed similar to a DAO, the applicant could decide how they would want to design the governance model:
- permissioned (invite only by consensus)
- open to token holders
- open to registered domains

Multiple discussions, especially in crypto-related conferences (recently @ UBRI), have been made to talk about the research, implementation and use of DAOs on the XRPL. Since smart contracts have yet to be implemented on the XRPL, using Evernode's HotPocket soluton, we're able to develop a layer-2 application that has a built-in DAO for its users.

### Decentralized Key Management 
To also stimulate the decentralized aspect of Evernode Applications, the service's XRPL account(s) should be in control of a list of its cluster's independent nodes via a signers list on the XRPL. This would essentially make the service trustless and decentralized as there's no centralized control over the application and its XRPL account(s), the selected nodes would be able to submit valid transactions to the XRPL by collecting their peers' signatures and submitting it to a rippled node for validation.

### XRPL Payments Integration
Since this is a wallet naming service, its users must pay for a domain name to the service's XRPL account. After payment has been made, the service would register the domain name and point it to the user's XRPL account and API service must be provided, on-chain and off-chain.

### API service
The wallet naming service should be able to point a domain name (.xrp) to a legitimate XRPL account via websocket (off-chain API via nodes) or XRPL transaction (on-chain API via service's XRPL account):

Forward record: request address @ `wojake999.xrp` -> return address `rf8kg7r5Fc8cCszGdD2jeUZt2FrgQd76BS`

Reverse record: request domain @ `rf8kg7r5Fc8cCszGdD2jeUZt2FrgQd76BS` -> return domain `wojake999.xrp`

# Evernode 101
## Evernode
Evernode is a layer-2 smart contract hosting platform, consisting of 3rd party servers (Evernode Hosts) providing hosting services to smart contract developers at scale.

Evernode as a hosting platform is trustless, open and affordable as a service. Evernode achieved its open and trustless element by utilizing the Hooks amendment to maintain a registry of hosts and a set of configurable parameters. This way, smart contract developers are able to go through the registry, get a list of hosts that would be apart of the application's cluster and acquire their hosting services through their respective Hosting NFTs. All of this happens with no oversight, approval or agreement from any party. It is decentralized and open for all, 24/7.

## HotPocket 
HotPocket is a [UNL](https://xrpl.org/faq.html#what-are-unique-node-lists-unls)-based consensus engine. Utilizing HotPocket, developers are able to deploy a smart contract onto a set of HotPocket nodes and construct a P2P decentralized network that receive user inputs, produce an output by executing the contract code, persist data, and perform consensus.

![image](https://user-images.githubusercontent.com/87929946/204559517-9be91e93-45e7-4f77-9a2f-cfc81f76722a.png)

## Helpful links/resources
Evernode Dev SDK: https://github.com/HotPocketDev/evernode-sdk

Evernode SDK tutorial: https://github.com/HotPocketDev/evernode-sdk#tutorials

Evernode Host: https://github.com/HotPocketDev/evernode-host

### Whitepapers
Evernode: https://lexautomagica.com/wp-content/uploads/2021/07/Evernode-Whitepaper-1.0.pdf

HotPocket: https://evernode.files.wordpress.com/2021/08/hotpocket-white-paper.pdf

Sashimono (a daemon): https://evernode.files.wordpress.com/2021/08/sashimono.pdf

### Notes
1. What is the high level explanation of this bounty?
> Research & Develop Evernode, its HotPocket solution and its capabilities to provide a plethora of layer-2 decentralized applications atop the XRPL. Build a wallet naming service using its HotPocket solution.
2. What problem is this solving?
> Introduce and open up the XRPL developer community to Evernode, introducing them to the possibilities of layer-2 dApps.
3. What are the requirements for this solution?
> Research, develop and open source the Wallet Naming Service's code on a GitHub repository dedicated for the community and explain how it works under the hood.
4. Why shouldn't this be applied as a Grant instead?
> This is an open research & development initiative on smart contracts & web3, it's intended to be of reference for the XRPL dev community to learn about Evernode, HotPocket and web3. This bounty should be able to be pursued by other developers as well. The XRPL needs more web3 talent and to incentive more developers to go down this road, a bounty would do the job for developers to research this area of expertise.
5. Does this help XRPL developers?
> Yes, the developer community and the broader community could refer to the initiative's materials and learn how Evernode, HotPocket, an L2 dApp works and a few web3 essentials like the governance system (DAO).
6. What skill sets will the applicant obtain during the process of the bounty and does it benefit the XRPL ecosystem?
> A plethora of skills from developing and coding the underlying code to designing the core features and how they operate with each other. One of the main skill sets that the applicant will be obtaining is the understanding of [Evernode](https://lexautomagica.com/wp-content/uploads/2021/07/Evernode-Whitepaper-1.0.pdf), [HotPocket](https://evernode.files.wordpress.com/2021/08/hotpocket-white-paper.pdf) and the [Developer SDK](https://github.com/HotPocketDev/evernode-sdk). The XRPL ecosystem needs more web3, infra, DeFi talent to progress forward. This initiative is not be limited to the understanding of Evernode because applicants learn about web3, infra and layer-2 solutions similar to Evernode (i.e. Optimus, zkSync) as they go down this path.

I am interested in this bounty, I will pursue it as I believe Evernode could provide the XRPL with the necessary applications to progress forward as the Internet of Value. Unfortunately, there's a very small number of research & development being done on Evernode. I've already worked on the application's basics, you can register a domain, retrieve an XRPL account via websocket. More features will be rolled out as time progresses.

Further more, this bounty's benefits is not limited to Evernode. The community could and should pick up this bounty to learn more about smart contracts and web3. The XRPL community is in dire need of more web3 talent, the ecosystem is growing with more and more job demands from independent projects around the world.

**Comment by JST5000** [comment]
Hey @wojake, thanks for sharing this proposal! 

XRPL Bounties has decided not to fund this bounty at this time. It requires research and does not appear to solve an immediate problem for the ecosystem, particularly because Evernode isn’t active until Hooks goes live.

You may want to consider applying to [XRPL Grants](https://xrplgrants.org/) when they begin accepting applications for the next Wave. XRPL Grants is intended for a project that you'd like to pick up yourself.

**Reply by wojake** [reply]
Hi, thank you for the update. I'll just continue building it myself then. In the mean time, I think there needs to be some clarification on my post:

While this initiative won't benefit the broader dev community, it will help a number of developers that are interested in developing on Evernode's HotPocket solution to some extend. Yes, Evernode isn't live until Hooks is enabled but it will help the ecosystem bootstrap a number of developers with knowledge on developing on HotPocket. Again, thank you ❤️

**Reply by xVet** [reply]
> Hey @wojake, thanks for sharing this proposal!
> 
> XRPL Bounties has decided not to fund this bounty at this time. It requires research and does not appear to solve an immediate problem for the ecosystem, particularly because Evernode isn’t active until Hooks goes live.
> 
> You may want to consider applying to [XRPL Grants](https://xrplgrants.org/) when they begin accepting applications for the next Wave. XRPL Grants is intended for a project that you'd like to pick up yourself.

Hi @JST5000 , 

i strongly believe this is a PoC we need rn in the community. There's a surge recently of questionable actors trying to profit of this exact idea with bad implementation and wrong concept. It clearly shows the demand of such a service in the community is here but it lacks a proper PoC and best practices that can be done now at the same time this is helping to tap into Hooks and to utilize solutions that are build with Hooks, like Evernode. 

I see this as a crucial part in enabling further exploration of Hooks, showcasing the real world possibilities with this amendment and to ultimately have the community build applications with Hooks and to apply potentially for the XRPL Grant, as a result of the work of this bounty.

With this, please reconsider your decision as i believe this bounty is a great addition and a benefit for us in the dev community.

Thank you very much!

**Reply by wojake** [reply]
Thank you @xVet ❤️ 

I've discussed about this matter and I think this initiative is a little too advanced for devs trying to understand how Evernode & HotPocket works, I've already finished coding the entire thing but I won't make it public. I'll close this proposal down and maybe propose something easier in the future that fits the Bounty program, where *all* devs that understand Evernode & HotPocket can take on.

An Introduction to Evernode & HotPocket with simple code snippets/programs on a repo that takes a dev step by step to understand how it all works. Your thoughts @JST5000 ?

**Reply by JST5000** [reply]
Hey WoJake, thanks for sharing the idea, but I don’t think it quite fits the bounty program for two main reasons.

1. We’ve already provided a grant of $200,000 to the Evernode project, so we’re wary of contributing more funds to the project.
2. We don’t want bounties for work that should be on the maintainer’s roadmap already.
    - Otherwise, we’re worried that instead of incentivising people to work on projects which are underserved, we’d be incentivising projects to wait on doing work until they check if they can get a bounty for it first.

Since the Evernode project is still going strong, the tutorial feels like something they should write or coordinate.

I agree that an accessible tutorial for Evernode would be useful, but a bounty doesn’t seem like the right way to make it happen.

**Reply by wojake** [reply]
As always, I appreciate the replies, the opportunities that were given to me in the Bounty program and the team's decision to not fund this bounty, it's reasonable 😄 . There is nothing even close to Evernode & HotPocket in crypto right now that's live, so ye, there's a lot that needs to built and said for it to be adopted.
